### DISCUSSION TITLE
User Email Verification and Registration

### DISCUSSION THREAD

**Discussion by james-d-elliott** [authelia/authelia#6093] [Ideas]
The idea is relatively simple (though has a few moving parts which affect security and is therefore unlikely to be a quick turnover, thus not a FR yet) and builds off of discussion #5198.

1. Record the email addresses the user has validated via one time code.
2. Allow end-users to individually setup a security-specific email via the UI for the purposes of all security-sensitive (or all) emails.

Critical Points:

- Emails used for validating something should only ever be sent to a single email.
- An attacker should not be able to see these emails in the UI in the clear.
- An attacker should not be able to pick which email a verification is delivered to in any way.

**Comment by ivanov17** [comment]
I will be glad to any solution to the issue of recovering user passwords. Sorry for the misunderstanding and thank you very much for your work.

**Comment by mrdrmmn** [comment]
Perhaps instead of providing the username for the reset, the end user could provide the email.  In the case an email address is provided, then that is the email address which authelia should send the reset link/code.

It is already possible to configure Authelia to all logging in using any email address assosicated with an account by setting the users_filter properly:

    ## To allow sign in both with username and email, one can use a filter like
    ## (&(|({username_attribute}={input})({mail_attribute}={input}))(objectClass=person))
 
 I have this set up and it work great.  This also already allow me to reset a password by providing an email address rather than the username.  The only issue is that authelia still send the email to the first email address set for the user it found.  But this could be changed fairly easily by simply checking specifically if the "input" is in the "mail_attribute" of the user and if so, send the reset to that email address.

I believe this would address this issue and be as safe a you possibly can be.  It probably still make sense to make it a configurable option as that should also be fairly trivial and ensure that the system administrator can decide if it is right for them or not independently of allowing logins using the email address.

**Comment by DoomHammer** [comment]
Here's my ugly attempt at getting at this: https://github.com/authelia/authelia/pull/10992

**Reply by Crowley723** [reply]
> Here's my ugly attempt at getting at this: https://github.com/authelia/authelia/pull/10992

The pr is a good start.

I believe the largest part of this feature will be the mechanism for "verifying" a user's email. As mentioned in the original post, the user has to verify emails with an otp/magic link sent to the email in question. Then, that email can be selected as the one to be used for password resets. Only verified emails should be selected.

**Reply by Crowley723** [reply]
Not sure how we want to handle multiple emails. It will probably look something like this:


- The user has multiple emails in the authentication backend.
- They can select one (after generating an elevated session) to verify, which will send an OTP or magic link email to the specified address.
- The address "verification" should be stored in the database.
- Verified addresses can be specified as the user's primary email for reset purposes. (Unsure how we want to handle this, if checking auth backend, then cross referencing with the database is sufficient)

**Reply by james-d-elliott** [reply]
I generally think if we facilitate a feature like this the user must enter it themselves and verify it and it's saved to the Authelia DB, i.e. distinctly separate from the Authentication Backend.

**Reply by DoomHammer** [reply]
@Crowley723 as far as I understand the email that comes from LDAP is not verified in either way. And it is the one that is used for recovery.

Maybe I'm missing something but I thought replacing one address with another with both coming from LDAP shouldn't make a difference and be "just as good"

**Reply by james-d-elliott** [reply]
It doesn't match the criteria outlined in the idea at all. If you'd like to propose a competing idea you may do so [here](https://github.com/authelia/authelia/discussions/new?category=ideas) and we can then ascertain which idea is more popular, and professionally debate the benefits/flaws of each . Thanks.
